src/System.Linq.Expressions/src/System/Runtime/CompilerServices/ReadOnlyCollectionBuilder.cs

```diff
@@ -131,7 +128,7 @@ public int IndexOf(T item)
         /// <param name="item">The object to insert into the <see cref="ReadOnlyCollectionBuilder{T}"/>.</param>
         public void Insert(int index, T item)
         {
-            if (index > _size)
+            if (index < 0 || index > _size)
```

Comment:
We should review #14059 and weigh exception compatibility against the (seemingly unintentional) inconsistency with other similar collection types such as `List<T>`. Who can make a call on these?

Of all the issues raised in #14059, the bug in `Reverse` should be addressed because it can expose the supposed-to-be-empty tail of the underlying array when out-of-range `index` and/or `count` values are passed.